GL3 attenuation for C++ raytracer

christophwelling opened the pull request:
This fixes the GL3 attenuation model for the C++ raytracer. Instead of using the GSL integrator (which had problems dealing with the GL3 model), the raytracer uses `get_path` to get points on the ray path, and then uses those to calculate the attenuation factor.
This is not 100% the same, but I tested the difference by generating a transmitter at random depths from 0.5m and 150m, and a transmitter at random depths between 30 and 3000m and distances between 50 and 3000m. This plot shows the difference between the attenuation factor of the C++ and the python raytracer, so the difference is under 1%.
![attenuation_test_random_points](https://user-images.githubusercontent.com/42930879/228969775-3aec02e4-eea7-4677-9744-a0be5f0c40d4.png)
To make it easier to compare, I also added the option to tell the raytracer to use python, even if the C++ implementation is available.

I also noticed another problem with the C++ raytracer: For frequencies above 1.4GHz or so, the interpolation can cause negative attenuation lenghts. Even though this is out of band, it can blow up the signal enough to overcome the filter. So I restricted the interpolation to 600MHz to prevent this.

cg-laser reviewed (changes requested):
thanks Christoph! Looks good. The feature to specify to use the python ray tracer is great. Did you change it att all occurrences? Maybe it would be better to only use `self._use_python_raytracer` which is set in the init function as 

`self._use_python_raytracer = use_python_raytracer or not cpp_available`

Could you also improve the log output so that it is logged that the python version is used? 

In this block in the beginning, could you change it to use the logger

```try:
    from NuRadioMC.SignalProp.CPPAnalyticRayTracing import wrapper
    cpp_available = True
    print("using CPP version of ray tracer")
except:
    print("trying to compile the CPP extension on-the-fly")
    try:
        import subprocess
        import os
        subprocess.call(os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                 "install.sh"))
        from NuRadioMC.SignalProp.CPPAnalyticRayTracing import wrapper
        cpp_available = True
        print("compilation was successful, using CPP version of ray tracer")
    except:
        print("compilation was not successful, using python version of ray tracer")
        print("check NuRadioMC/NuRadioMC/SignalProp/CPPAnalyticRayTracing for manual compilation")
        cpp_available = False
```
Here I would change the log output to "CPP ray tracer available. It will be used if not specified otherwise".

christophwelling commented:
Okay, I changed the output to use the logger and give clearer feedback.
I changed it in all places where there is a check if the C++ raytracer is available, so that should be all of them.

christophwelling commented:
@cg-laser Can you approve this?

colemanalan commented:
The GL3 attenuation model is not working correctly. After running some simulations using `greenland_simple` as the icemodel and GL3. I get the following results for the distances as which triggers occur. Most importantly, once reaching 10^17eV, the distributions "saturate".

![image](https://github.com/nu-radio/NuRadioMC/assets/74733466/50687740-05d2-47cc-9bc5-ed7d18341516)

Changing only `GL3` --> `GL2` in my configuration file produced results that looked more like I would expect (no saturation).

![image](https://github.com/nu-radio/NuRadioMC/assets/74733466/da8cae3e-5c08-4665-b6d0-b17756864e57)

christophwelling commented:
I finally found the time to look into this, and I have no idea how this can be happening. I compared the attenuation factors the C++ raytracer returns for the GL1, GL2 and GL3 models when raytracing between different points, and they were always relatively similar. Of course there were differences between them, since they are different models, but nothing that could explain Alan's plots (especially since GL3 usually returned a slightly larger signals than GL2). The only explanation I can think of is that there are events deeper than 3km, i.e. below the bedrock, where the models do return very different values. But Alan assured me this is not the case. So I cannot reproduce this! Is there a way I could see the simulations that showed this strange behavior?

fschlueter commented:
I merged develop and #918 into this branch to be able to compare the calculations.

fschlueter commented:
@christophwelling why did you implement this cap for the frequency? I am not getting any negative attenuation length with GL3:
<img width="1259" alt="image" src="https://github.com/user-attachments/assets/dc4a600f-f533-422d-b8ce-1adb036086d4" />

fschlueter commented:
Same plots as in #918 but now with the optimised calculation for both the cpp and python wrapper. The cpp wrapper now takes again the crown for the fasters implementation thanks to @christophwelling. Results are comparable I would say
<img width="897" alt="image" src="https://github.com/user-attachments/assets/d209c4f0-cfcb-43a6-91a6-ed80b84f06a2" />

<img width="897" alt="image" src="https://github.com/user-attachments/assets/ed243f19-d64c-4330-bfd9-6172b862d6d1" />

<img width="897" alt="image" src="https://github.com/user-attachments/assets/e4a0910e-cb10-430d-a928-b261bac30183" />

fschlueter commented:
next I think I will try to run neutrino simulations with both implementations to see whether they differ.

fschlueter commented:
@cg-laser, @sjoerd-bouma, @christophwelling  any comment yet? Do you think the shown accuracy is okay? Two things I still have in mind are:
- Checking whether the attenuation (as a function of the frequency) change for a given source but different receivers. 
- Running actual simulations and recalculating Veff.

fschlueter commented:
These are vertex distributions for triggered events with GL1 and GL3 (not the same events at generator level). I think they look fine. What @philippwindischhofer noticed is that for GL3 there are a events missing at very small distances to the antenna. I checked the attenuation calculation for very small distances but could not yet spot a problem.
![distance_bins_GL1](https://github.com/user-attachments/assets/d56f65f9-fa6a-4da8-9a44-e7811c8b70fc)
![distance_bins_GL3](https://github.com/user-attachments/assets/1e444b29-ff59-4a5d-880a-20d37fdc8297)
![vertex_distr_GL1](https://github.com/user-attachments/assets/236390a4-0fe3-48cb-a0d2-7960705e5030)
![vertex_distr_GL3](https://github.com/user-attachments/assets/91bbff5d-5c27-4c33-acf4-fae46b2dbcc4)

anelles commented:
As far as I can see, the distributions look ok now. I think a lot has changed since Alan first reported this. So I would say we don't necessarily need to do archeology to retrieve the cause of the issue.

fschlueter commented:
This is the comparison between the calculation from the PR "cpp" and from #918 "cpp with GSL". This calculation is much faster! I guess we are sticking with this one?

![image](https://github.com/user-attachments/assets/354a8c79-de38-4fd4-ab37-dc40f126a39a)
![distance_bins_GL3_cpp](https://github.com/user-attachments/assets/889a09e9-c616-4b45-837e-d7c53513cc0c)
![vertex_distr_GL3_cpp](https://github.com/user-attachments/assets/d0a13848-ba7a-4e28-adad-36ca6956a126)
![distance_bins_GL3_-_cpp_with_GSL](https://github.com/user-attachments/assets/1ec1d0f1-2e8b-468c-bcfa-b51c390bf45b)
![vertex_distr_GL3_-_cpp_with_GSL](https://github.com/user-attachments/assets/40e8dd18-c6e9-4eda-a9b8-65dbdf93e8b5)

fschlueter commented:
I merged PR #918 into this PR which makes the diff convoluted sorry. But in any case @christophwelling or @sjoerd-bouma are you happy with that and the cross-checks I made?

anelles commented:
In concur! Thanks for getting this implemented everyone!

sjoerd-bouma commented:
Actually... I've just remembered I've thought about this problem before, and I'm now wondering if the 'smarter' way to do this wouldn't be to normalize the result returned by `get_attenuation_integral_GL3` by multiplying it by `analytic_path_length / segmented_path_length`; this should be essentially instantaneous, and I'm not sure what the order of the error is in splitting up the path into 100 segments and computing the total path length... but I guess it could be of the order of 1% for some refracted/reflected paths with a lot of curvature?

Although I'm not sure if the analytic path length is in the C++ code already or it would first need to be ported from the Python version. If the latter, we can defer this to a future PR.

fschlueter commented:
Great, I will merge this PR asap and see what remains relevant in #918. @sjoerd-bouma I am not sure about the relevance of what you propose. But it sounds like it can be implemented easily. Nonetheless I would like to refer it to a future PR (if you have some capacity for it)
